File: src/libraidzero/controller/pidController.cpp
#include "libraidzero/controller/pidController.hpp"
#include "okapi/api/units/QTime.hpp"
#include <algorithm>
#include <cmath>

PIDController::PIDController(double kP, double kI, double kD) : 
    kP{kP}, kI{kI}, kD{kD} 
{

}

void PIDController::setGains(double p, double i, double d) {
    kP = p;
    kI = i;
    kD = d;
}

void PIDController::setOutputLimits(double min, double max) {
    minOutput = min;
    maxOutput = max;
}

void PIDController::setSetpoint(double sp) {
    setpoint = sp;
}

double PIDController::getSetpoint() {
    return setpoint;
}

void PIDController::setTolerance(double positionTolerance, double velocityTolerance) {
    posTol = positionTolerance;
    velTol = velocityTolerance;
}

bool PIDController::atSetpoint() {
    return std::abs(positionError) < posTol && std::abs(velocityError) < velTol; 
}

double PIDController::calculate(double measurement, double sp) {
    setSetpoint(sp);
    return calculate(measurement);
}

double PIDController::calculate(double measurement) {
    previousError = positionError;

    positionError = setpoint - measurement;
    velocityError = (positionError - previousError) / timer->getDt().convert(okapi::second);

    totalError += positionError;
    totalError = std::clamp(totalError, minOutput, maxOutput);

    return std::clamp(kP * positionError + kI * totalError + kD * velocityError, minOutput, maxOutput);
}

double PIDController::getPositionError() {
    return positionError;
}

double PIDController::getVelocityError() {
    return velocityError;
}

void PIDController::reset() {
    previousError = 0.0;
    totalError = 0.0;
}

File: visualization/MecanumStrafe/PIDController.java
public class PIDController {
    
    public static class Gains {
        float kP, kI, kD;
        
        public Gains(float kP, float kI, float kD) {
            this.kP = kP;
            this.kI = kI;
            this.kD = kD;
        }
    }
    
    private float kP, kI, kD;
    private float setpoint;
    
    private float posTol = 3.0f;
    private float velTol = Float.POSITIVE_INFINITY;
    
    private float previousError, totalError, positionError, velocityError;
    
    public float clamp(float val, float min, float max) {
        return Math.max(min, Math.min(max, val));
    }

    public PIDController(Gains gains) {
        kP = gains.kP;
        kI = gains.kI;
        kD = gains.kD;
    }
    
    public void setSetpoint(float setpoint) {
        this.setpoint = setpoint;
    }
    
    public boolean atSetpoint() {
        return Math.abs(positionError) < posTol && Math.abs(velocityError) < velTol; 
    }
    
    public float calculate(float measurement, float dt) {
        previousError = positionError;

        positionError = setpoint - measurement;
        velocityError = (positionError - previousError) / dt;
    
        totalError += positionError;
        totalError = clamp(totalError, -1.0f, 1.0f);
    
        return clamp(kP * positionError + kI * totalError + kD * velocityError, -1.0f, 1.0f);
    }
    
    public void reset() {
        previousError = 0.0f;
        positionError = 0.0f;
        velocityError = 0.0f;
        totalError = 0.0f;
    }
}
